backend/src/apiserver/server/util.go

```diff
@@ -216,64 +216,88 @@ func ValidateExperimentResourceReference(resourceManager *resource.ResourceManag
 	return nil
 }
 
-func ValidatePipelineSpec(resourceManager *resource.ResourceManager, spec *api.PipelineSpec) error {
-	if spec == nil || (spec.GetPipelineId() == "" && spec.GetWorkflowManifest() == "") {
-		return util.NewInvalidInputError("Please specify a pipeline by providing a pipeline ID or workflow manifest.")
+func ValidatePipelineSpecAndResourceReferences(resourceManager *resource.ResourceManager, spec *api.PipelineSpec, resourceReferences []*api.ResourceReference) error {
+	// centralize all the nil checks here
+	pipelineId := spec.GetPipelineId()
+	workflowManifest := spec.GetWorkflowManifest()
+	pipelineVersionId := getPipelineVersionIdFromResourceReferences(resourceManager, resourceReferences)
+
+	if err := validateParameters(spec.GetParameters()); err != nil {
+		return err
+	}
+	if workflowManifest != "" {
+		if pipelineId != "" || pipelineVersionId != "" {
+			return util.NewInvalidInputError("Please don't specify a pipeline version or pipeline ID when you specify a workflow manifest.")
+		}
+		if err := validateWorkflowManifest(spec.GetWorkflowManifest()); err != nil {
+			return err
+		}
+	} else {
+		if pipelineId == "" && pipelineVersionId == "" {
+			return util.NewInvalidInputError("Please specify a pipeline by providing a pipeline spec or/and pipeline version.")
```

Comment:
nit: or workflow manifest

Reply:
The reason I use pipeline spec here instead of workflow manifest and pipeline id is because pipeline id can coexist with pipelineversion but workflow manifest can't.

Reply:
+1, I like your rephrased version

Reply:
done